### PR TITLE
explicit signature scheme and hash algorithm

### PR DESCRIPTION
**Pull request by rautammi**
At least my Infineon OPTIGA SLB9670 requires signature scheme and hash algorithm to be stated explicitly.

### CODE REVIEW

**Comment by rautammi** [comment]
I signed it!

**Comment by awly** [comment]
@rautammi is the email you used for commit (`rauno.tamminen@insta.fi`) same as the one you used to sign the CLA and used for your GitHub account?

File: tpm2/tpm2.go [hunk 1]
```diff
@@ -985,7 +985,7 @@ func encodeSign(key tpmutil.Handle, password string, digest []byte) ([]byte, err
 	if err != nil {
 		return nil, err
 	}
-	params, err := tpmutil.Pack(digest, AlgNull, tagHashCheck, HandleNull, []byte(nil))
+	params, err := tpmutil.Pack(digest, sigScheme, hashAlg, tagHashCheck, HandleNull, []byte(nil))
```
Comment (awly) [inline tpm2/tpm2.go]: This only works for RSA, HMAC and some ECC signing schemes.
Anonymous ECC schemes also take a count (see section 11.2.1.3 in https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf).

Ideally, switch on `sigScheme`:
- for `AlgNull` do nothing
- for `AlgECDAA` add hash and count
- for the rest add a hash

Comment (rautammi) [reply tpm2/tpm2.go]: So, sigScheme should be an interface, like crypto.SignerOpts, which implements HashFunc() and , in case of AlgECDAA, also Count() ?

Comment (awly) [reply tpm2/tpm2.go]: If you want to abstract it, make it like https://github.com/google/go-tpm/blob/master/tpm2/structures.go#L72
Have a dedicated type and switch within to decide how to encode.

Or just add a switch here. Up to you.

**Comment by rautammi** [comment]
My bad with messing up with author email addresses. I may have to close this pull request and create a new one with correct email addresses.

But now there's the switch added. I am not convenient with bloating the function interface with seldom needed parameters, but I am lacking a vision of a really elegant solution.

File: tpm2/tpm2.go [hunk 2]
```diff
@@ -1003,8 +1013,8 @@ func decodeSign(buf []byte) (*Signature, error) {
 
 // Sign computes a signature for digest using a given loaded key. Signature
 // algorithm depends on the key type.
-func Sign(rw io.ReadWriter, key tpmutil.Handle, password string, digest []byte) (*Signature, error) {
-	cmd, err := encodeSign(key, password, digest)
+func Sign(rw io.ReadWriter, key tpmutil.Handle, password string, digest []byte, sigScheme, hashAlg Algorithm, count uint32) (*Signature, error) {
```
Comment (awly) [inline tpm2/tpm2.go]: Ah, I totally forgot, `TPMT_SIG_SCHEME` is already (almost) implemented: https://github.com/google/go-tpm/blob/master/tpm2/structures.go#L256

Could you add `count` field (and handling of ECDAA) to it instead and use a `*SigScheme` argument for `Sign`?

**Comment by awly** [comment]
To fix the email you can squash your commits instead and use the correct email there.
Something like
```
git reset --soft 4e28e10ca34602b0156c61a478b61499bce142d5
git config --global user.email "email@example.com"
git add .
git commit
git push --force
```

File: tpm2/tpm2.go (hunk 2 above)
Comment (tmroeder) [reply tpm2/tpm2.go]: I agree. Using SigScheme in Sign sounds like the right choice for the API.

File: tpm2/tpm2.go [hunk 3]
```diff
@@ -976,7 +976,11 @@ func Shutdown(rw io.ReadWriter, typ StartupType) error {
 	return err
 }
 
-func encodeSign(key tpmutil.Handle, password string, digest []byte, sigScheme, hashAlg Algorithm, count uint32) ([]byte, error) {
+func encodeSign(key tpmutil.Handle, password string, digest []byte, sigScheme *SigScheme) ([]byte, error) {
+	if sigScheme == nil {
```
Comment (awly) [inline tpm2/tpm2.go]: This shouldn't be necessary. `SigScheme.encode` handles nil

File: tpm2/tpm2.go [hunk 4]
```diff
@@ -976,7 +976,11 @@ func Shutdown(rw io.ReadWriter, typ StartupType) error {
 	return err
 }
 
-func encodeSign(key tpmutil.Handle, password string, digest []byte) ([]byte, error) {
+func encodeSign(key tpmutil.Handle, password string, digest []byte, sigScheme *SigScheme) ([]byte, error) {
+	if sigScheme == nil {
```
Comment (awly) [inline tpm2/tpm2.go]: Remove this check, it's handled by SigScheme.encode

**Comment by awly** [comment]
Thanks for fixing the CLA. One last comment and LGTM

**Pull request merged**